#14 redo deal screen Part 1

ValCanBuild opened the pull request:
This PR got out of hand because I've tried to implement a Reactive architecture approach for the DeaListActivity. That whole screen is still partly in progress - it'll be fixed in my next PR. I wanted to submit this though since the code is already so big and there's a lot of other non-ui changes that I've created that need to be addressed

Commit:
- Renaming top_navigation package to deals package
- Creating DealListAdapter and StoreImageAdapter
- Creating DealEntryUiModel to represent an item in the new adapter
- Introducing some structure of new UI - renaming TopNavigationActivity to DealListActivity
- Adding spinner to DealListActivity which will control what deals are seen
- Creating new GameDeal object

Commit:
- Updating StoreFilter to provide an activeStoreIds observable

Commit:
- Creating GroupedGameDeal

Commit:
- Putting structures in place for new Deal Fetching.

Commit:
- First shot at DB caching for deals

Commit:
- Cleaning up

Commit:
- DealListActivity working with multiple sort types

Commit:
Fixing tests

Commit:
First shot at JakeWharton style reactive architecture for DealList

Commit:
- Adding loading state to DealListActivity
- Fully reactive UI

Commit:
- Fixing toolbar styling

Commit:
- Add bulgarian translations so lint stops complaining

Commit:
All DealListActivity events are now part of the stream

Commit:
- Working version of DealListActivity with new Reactive architecture.
- Limitation - deal list is always reloaded when the store filter is changed
- No interactions with Deal Objects yet

MungoRae commented on app/src/main/kotlin/com/rockspin/bargainbits/data/models/GroupedGameDeal.kt:
```diff
@@ -0,0 +1,30 @@
+package com.rockspin.bargainbits.data.models
+
+import java.util.*
```
I don't know if Kotlin is different but in Java this would mean importing a whole load of classes you are not using.

ValCanBuild replied:
This is super weird - the Kotlin auto complete includes the whole util package, not just Date. I can manually fix it in this case but i don't think i'll pay attention when I auto import in future.

MungoRae commented on app/src/main/kotlin/com/rockspin/bargainbits/data/models/GroupedGameDeal.kt:
```diff
@@ -0,0 +1,30 @@
+    val dealRating = deals.maxBy { it.dealRating }!!.dealRating
+    val thumbUrl = deals[0].thumbUrl
+    val releaseDate by lazy {
+        Date(deals[0].releaseTimestampSeconds * 1000)
```
Being really strict I could say this was a magic number.

MungoRae commented on app/src/main/kotlin/com/rockspin/bargainbits/data/models/daos/DbDealsDao.kt:
```diff
@@ -0,0 +1,21 @@
+import io.reactivex.Flowable
+
+/**
+ * Created by valentin.hinov on 29/05/2017.
```
These comments gives warning in Android Studio. You should not 'own' any code and the date isn't really that useful, see version control instead. Remove this comment in this file and the other files.

MungoRae commented on app/src/main/kotlin/com/rockspin/bargainbits/data/models/daos/GameDealDao.kt:
```diff
@@ -0,0 +1,21 @@
+ * Created by valentin.hinov on 29/05/2017.
+ */
+@Dao
+interface GameDealDao {
```
The previous class was called `DbDealsDao`, whereas this one omits the `Db`, this seems like inconsistent naming. Can we either have `Db` for every `Dao` class or remove it entirely. I would prefer to remove it as I hate seeing abbreviations in names.

ValCanBuild replied:
The other class is named `DbDealsDao` because it uses `DbDeals` objects. I agree the DbDeals name is not great but I was struggling to figure out what to name it

MungoRae commented on app/src/main/kotlin/com/rockspin/bargainbits/data/repository/deals/BBGameDealRepository.kt:
```diff
@@ -0,0 +1,65 @@
+/**
+ * Created by valentin.hinov on 29/05/2017.
+ */
+class BBGameDealRepository(
```
Unless there will be more than one GameDealRepository I don't see the need to prefix it with `BB`. This is the BargainBytes app.

MungoRae replied:
Ah I have just realised you called the interface GameDealRepository. I still hate initials but ok.

ValCanBuild replied:
I agree in principle - I inherited the "BB" prefix from what we had in the code before. I'm open to changing it after but for now all repository-like implementation objects should have it for consistency sake.

MungoRae commented on app/src/main/kotlin/com/rockspin/bargainbits/data/repository/deals/BBGameDealRepository.kt:
```diff
@@ -0,0 +1,65 @@
+
+        val apiCall = apiCallForParams(sortType, sortedIds)
+
+        // TODO - introduce DB cache "freshness" so cached data is not fetched if too old
```
Does this have an issue to track it? TODOs will never get done unless you do them or have something that bugs you until you have done them.

ValCanBuild replied:
Noted and created: https://github.com/Rockspin/BargainBytesAndroid/issues/25

MungoRae commented on app/src/main/kotlin/com/rockspin/bargainbits/data/repository/deals/BBGameDealRepository.kt:
```diff
@@ -0,0 +1,65 @@
+            .toObservable()
+    }
+
+    // TODO - Action that cleanups unreferenced GameDeals in database based on DbDeals
```
Again can this have an issue that tracks this TODO?

ValCanBuild replied:
Noted and created https://github.com/Rockspin/BargainBytesAndroid/issues/26

MungoRae commented on app/src/main/kotlin/com/rockspin/bargainbits/data/rest_client/GameApiService.kt:
```diff
@@ -20,4 +22,7 @@ interface GameApiService {
     fun getStores(): Single<List<Store>>
+
+    @GET("deals")
+    fun getDeals(@QueryMap requestParams: Map<String, String>): Single<List<GameDeal>>
```
Can we not use a QueryMap here? I find it is easier to read exactly what API calls are being made if we explicitly say what params are passed in. I would say that is the main advantage of using a Retofit interface.

MungoRae commented on app/src/main/kotlin/com/rockspin/bargainbits/data/repository/deals/BBGameDealRepository.kt:
```diff
@@ -0,0 +1,65 @@
+
+    private fun apiCallForParams(sortType: DealSortType, storeIds: Set<String>): Observable<List<GameDeal>> {
+        val params = mapOf(
+            "onSale" to "1",
```
These should really be in the retrofit interface.

ValCanBuild replied:
I can't put this in the retrofit instance since retrofit creates an implementation of the `GameApiService`. I would need to create another abstraction on top that uses GameApiService internally so that I can pass the parameters in a more "clean" way. It just seemed like too much unnecessary abstraction. Is it that big of an issue to have this here?

MungoRae commented on app/src/main/kotlin/com/rockspin/bargainbits/data/repository/deals/BBGameDealRepository.kt:
```diff
@@ -0,0 +1,65 @@
+
+    private fun apiCallForParams(sortType: DealSortType, storeIds: Set<String>): Observable<List<GameDeal>> {
+        val params = mapOf(
+            "onSale" to "1",
```
Also `"1"` is a magic number here

ValCanBuild replied:
Not really, "1" is true since the API uses 1 and 0 as true and false

MungoRae commented on app/src/main/kotlin/com/rockspin/bargainbits/ui/deals/DealListActivity.kt:
```diff
@@ -0,0 +1,136 @@
+        val viewModel = ViewModelProviders.of(this, factory).get(DealListViewModel::class.java)
+
+        listOf(
+            binding.dealSortTypeSpinner.itemSelections().map { DealListEvent.SortingChanged(it) }.skip(1),
```
It's great that with Kotlin we can now do 500 things in one statement on one line but that doesn't mean we _have_ to. This is very hard to read as it has many chains inside of one big chain. Could this be split up into several methods or variables?

MungoRae replied:
For example: In Jake Wharton's videos he would break up a big observables into sections like reacting to state, handling data and calling back to UI, and also for each event. https://youtu.be/0IKHxjkgop4?t=31m30s. Maybe that would help here.

ValCanBuild replied:
I'm still trying to tie down how this architecture ideally looks. Could you please ignore this part for now and keep your notes in mind for further PRs? The comments are valid but I'd like to have another review once i've cleaned this up.

MungoRae commented on app/src/main/kotlin/com/rockspin/bargainbits/ui/deals/DealListActivity.kt:
```diff
@@ -0,0 +1,136 @@
+                    DealListUiState.Navigation.SEND_FEEDBACK -> sendFeedbackEmail()
+                }
+
+            }.addTo(onDestroyDisposable)
```
I'm getting `hasError = true` but you don't seem to handle it, so I just get a blank screen.

MungoRae replied:
Also the error is not logged anywhere

ValCanBuild replied:
Like I said in the previous comment - this screen and architecture are still in a bit of a flux. Will handle better in another PR

MungoRae replied:
Ok, but I feel that everything going into develop should be releasable, and if you don't handle errors then you would be giving users bad experiences when the API was down or failed.

ValCanBuild replied:
I agree in principle but we know that develop at the moment is not releasable so we can skew the rules in this case.

MungoRae reviewed (commented):
I have gotten about half way through this. Couldn't get any data from the server at the moment, not sure why. Need to look into it further but I'm still getting used to this combination of Rx and Kotlin. 

Left some stuff for you to chew on in the mean time.

Commit:
PR Review
- Magic Number fixes
- Remove auto gen class headers
- Create "DealQuery" object to handle generating the parameter map for the API

MungoRae commented on app/src/main/kotlin/com/rockspin/bargainbits/data/repository/deals/BBGameDealRepository.kt:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    private fun apiCallForParams(sortType: DealSortType, storeIds: Set<String>): Observable<List<GameDeal>> {
+        val dealQuery = DealQuery(onSale = true, sortBy = sortType, storeIds = storeIds)
```
I don't know why this make it any better, if fact I feel this makes it worse. Now you have a whole class to do something that could be done in one line of an interface. It's like your programming for more modifications but you have none. YAGNI. But I have said this already so fine.

ValCanBuild replied:
![Yagni](https://s3-eu-west-1.amazonaws.com/productiveprogrammer/wp-content/uploads/2016/03/25205331/yagni.jpg)

Fair point, will change

MungoRae commented on app/src/main/kotlin/com/rockspin/bargainbits/ui/deals/DealListActivity.kt:
```diff
@@ -0,0 +1,136 @@
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        // TODO - remember user default choice for deal sorting
```
Another TODO here that will still be here in six months if it isn't tracked properly.

ValCanBuild replied:
https://github.com/Rockspin/BargainBytesAndroid/issues/27

MungoRae commented on app/src/main/kotlin/com/rockspin/bargainbits/ui/deals/DealListActivity.kt:
```diff
@@ -0,0 +1,136 @@
+        shareIntent.action = Intent.ACTION_SEND
+        shareIntent.type = "text/plain"
+        shareIntent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.share_app_subject))
+        val shareText = getString(R.string.share_app_text, "http://play.google.com/store/apps/details?id=" + packageName)
```
magic string of URL feels like it shouldn't just be hidden here.

MungoRae commented on app/src/main/kotlin/com/rockspin/bargainbits/ui/deals/DealListViewModel.kt:
```diff
@@ -0,0 +1,149 @@
+import io.reactivex.rxkotlin.merge
+import io.reactivex.rxkotlin.ofType
+import java.text.SimpleDateFormat
+import java.util.*
```
could be optimised

MungoRae reviewed (approved):
I have left some more comments, but I'll approve and let you decide if you want to do anything with them.

Commit:
PR review notes
